From be9cd0d123dfc1e15cd59f103918122f2e0fff53 Mon Sep 17 00:00:00 2001
From: Yannick Fertre <yannick.fertre@foss.st.com>
Date: Thu, 29 Feb 2024 11:09:09 +0100
Subject: [PATCH 0946/1141] drm/stm: set max segment size

Configure the DMA segment size to make sure we get
contiguous & coherent memory for framebuffer allocation.

Change-Id: Ie3ed9349b2d7f10d4613993b2b2e21e7d84379f1
Signed-off-by: Yannick Fertre <yannick.fertre@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/363732
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/gpu/drm/stm/drv.c | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/stm/drv.c b/drivers/gpu/drm/stm/drv.c
index 4700a9331677..06f7cc1171a9 100644
--- a/drivers/gpu/drm/stm/drv.c
+++ b/drivers/gpu/drm/stm/drv.c
@@ -223,7 +223,18 @@ static int stm_drm_platform_probe(struct platform_device *pdev)
 	if (ret)
 		goto err_suspend;
 
-	dma_set_coherent_mask(dev, DMA_BIT_MASK(32));
+	/* Configure the DMA segment size to make sure we get contiguous & coherent memory */
+	ret = dma_set_coherent_mask(dev, DMA_BIT_MASK(32));
+	if (ret) {
+		dev_err(dev, "Failed to set DMA segment mask\n");
+		goto err_suspend;
+	}
+
+	ret = dma_set_max_seg_size(dev, DMA_BIT_MASK(32));
+	if (ret) {
+		dev_err(dev, "Failed to set DMA segment size\n");
+		goto err_suspend;
+	}
 
 	ddev = drm_dev_alloc(&drv_driver, dev);
 	if (IS_ERR(ddev)) {
-- 
2.39.2

